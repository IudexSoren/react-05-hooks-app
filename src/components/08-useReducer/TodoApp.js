import { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';

import TodoList from './TodoList';
import TodoAdd from './TodoAdd';

import './styles.css';


const init = () => {
  return JSON.parse(localStorage.getItem('todos')) || [];
}

const TodoApp = () => {

  const [todos, dispatch] = useReducer(todoReducer, [], init);

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const handleDelete = ( todoId ) => {
    const action = {
      type: 'delete',
      payload: todoId
    }
    dispatch(action);
  }

  const handleComplete = (todoId) => {
    const action = {
      type: 'complete',
      payload: todoId
    }
    dispatch(action);
  }

  const handleAddTodo = (newTodo) => {
    const action = {
      type: 'add',
      payload: newTodo
    }
    dispatch(action);
  }

  return (
    <div>
      <h1>TodoApp ({ todos.filter(t => !t.done).length } pendientes)</h1>
      <hr/>
      <div className="row">
        <div className="col-4">
          <TodoAdd
            handleAddTodo={ handleAddTodo }
          />
        </div>
        <div className="col-8">
          Todos
          <TodoList
            todos={ todos }
            handleComplete={ handleComplete }
            handleDelete={ handleDelete }
          />
        </div>
      </div>
    </div>
  )
}

export default TodoApp
