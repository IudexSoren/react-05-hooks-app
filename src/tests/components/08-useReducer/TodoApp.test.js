import { shallow, mount } from 'enzyme';
import { act } from '@testing-library/react';
import TodoApp from '../../../components/08-useReducer/TodoApp';
import { initialState } from '../../fixtures/demoTodos';

describe('Pruebas de TodoApp', () => {

  const wrapper = shallow(<TodoApp />);
  Storage.prototype.setItem = jest.fn();

  test('Debe renderizar correctamente el componente', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('Debe agregar un Todo', () => {
    const wrapper = mount(<TodoApp />);
    act(() => {
      wrapper.find('TodoAdd').prop('handleAddTodo')(initialState[0]);
      wrapper.find('TodoAdd').prop('handleAddTodo')(initialState[1]);
    });

    expect(wrapper.find('h1').text()).toBe('TodoApp (2 pendientes)');
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);
  });

  test('Debe eliminar un Todo', () => {
    wrapper.find('TodoAdd').prop('handleAddTodo')(initialState[0]);
    wrapper.find('TodoList').prop('handleDelete')(initialState[0].id);

    expect(wrapper.find('h1').text()).toBe('TodoApp (0 pendientes)');
  });


})
