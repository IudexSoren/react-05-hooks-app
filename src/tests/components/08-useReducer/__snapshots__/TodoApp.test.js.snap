// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de TodoApp Debe renderizar correctamente el componente 1`] = `
<div>
  <h1>
    TodoApp (
    0
     pendientes)
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-4"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
    <div
      className="col-8"
    >
      Todos
      <TodoList
        handleComplete={[Function]}
        handleDelete={[Function]}
        todos={Array []}
      />
    </div>
  </div>
</div>
`;
